packages/eslint-plugin-react-hooks/src/ExhaustiveDeps.js

```diff
@@ -1381,50 +1399,97 @@ function collectRecommendations({
   };
 }
 
-// Finds functions declared as dependencies
+// If the node will result in constructing a referentially unique value, return
+// its human readable type name, else return null.
+function getConstructionExpresionType(node) {
+  switch (node.type) {
+    case 'ObjectExpression':
+      return 'object';
+    case 'ArrayExpression':
+      return 'array';
+    case 'ArrowFunctionExpression':
+    case 'FunctionExpression':
+      return 'function';
+    case 'ClassExpression':
+      return 'class';
+    case 'ConditionalExpression':
+      if (
+        getConstructionExpresionType(node.consequent) != null ||
+        getConstructionExpresionType(node.alternate) != null
+      ) {
+        return 'conditional';
+      }
+      return null;
+    case 'LogicalExpression':
+      if (
+        getConstructionExpresionType(node.left) != null ||
+        getConstructionExpresionType(node.right) != null
+      ) {
+        return 'logical expression';
+      }
+      return null;
+  }
```

Comment:
I believe that you could also add support for JSX here, because under the hood they are going to create a new object ref each render (right? I don't know the exact semantics of JSX transforms)

```js
const frag = <></>; // JSXFragment
const comp = <div />; // JSXElement
useMemo(() => {}, [frag, comp]);
```

Reply:
could also add handling for `AssignmentExpression`:
```js
let x = null;
let y = x = () => {};
useMemo(() => {}, [y]);
```

(i.e. ignore the `left` of the assignment)

Reply:
could also add handling for `NewExpression`?
```js
const x = new Map();
useMemo(() => {}, [x]);
```

Reply:
could also add handling for `Literal` when `value instanceof RegExp`?
```js
const x = /a/;
useMemo(() => {}, [x]);
```

Reply:
could also handle TS type assertions (`TSAsExpression`):

```js
const x = {} as any;
useMemo(() => {}, [x]);
```

(i.e. ignore the node and check the `expression`.)

Reply:
could also handle flow type casts (`TypeCastExpression`):

```js
const x = ({}: any);
useMemo(() => {}, [x]);
```

(i.e. ignore the node and check the `expression`.)

Reply:
This list is gold. Thanks! I had intentionally omitted `NewExpression` thinking that `String` `Number` and `Boolean` would be compared by value, but I forgot that they will be boxed so presumably they _will_ be referentially unique and therefore worth flagging.

Reply:
Yeah I didn't think about that, but testing in chrome shows they are referentially unique
![image](https://user-images.githubusercontent.com/7462525/90067287-bef92380-dca3-11ea-9ebf-c3eda03fd88e.png)

Note that they're also not even loosely equal:

![image](https://user-images.githubusercontent.com/7462525/90067386-e18b3c80-dca3-11ea-83dc-ed64fee3bb78.png)

(note: if you swap one side for a literal, it coerces the object to the same type and then they become loosely equal)

Reply:
This is such an edge case btw. I don't think handling boxed primitives is super important.

Reply:
@gaearon I would agree that it's very unlikely to happen, we get it for free when handling other `NewExpression`s like `new Map([])`.